libs/vkd3d/swapchain.c

```diff
@@ -0,0 +1,2207 @@
+    barrier.dstQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED;
+    barrier.image = image;
+    barrier.subresourceRange.aspectMask = VK_IMAGE_ASPECT_COLOR_BIT;
+    barrier.subresourceRange.baseMipLevel = 0;
+    barrier.subresourceRange.levelCount = VK_REMAINING_MIP_LEVELS;
+    barrier.subresourceRange.baseArrayLayer = 0;
+    barrier.subresourceRange.layerCount = VK_REMAINING_ARRAY_LAYERS;
+
+    vk_procs->vkCmdPipelineBarrier(cmd_buffer,
+            src_stage_mask, dst_stage_mask, 0, 0, NULL, 0, NULL, 1, &barrier);
+}
+
+static VkResult d3d12_swapchain_record_swapchain_blit(struct d3d12_swapchain *swapchain,
+        VkCommandBuffer vk_cmd_buffer, VkImage vk_dst_image, VkImage vk_src_image)
+{
+    const struct vkd3d_vk_device_procs *vk_procs = d3d12_swapchain_procs(swapchain);
+    VkCommandBufferBeginInfo begin_info;
+    VkImageBlit blit;
+    VkFilter filter;
+    VkResult vr;
+
+    if (swapchain->desc.Width != swapchain->vk_swapchain_width
+            || swapchain->desc.Height != swapchain->vk_swapchain_height)
+        filter = VK_FILTER_LINEAR;
+    else
+        filter = VK_FILTER_NEAREST;
+
+    begin_info.sType = VK_STRUCTURE_TYPE_COMMAND_BUFFER_BEGIN_INFO;
+    begin_info.pNext = NULL;
+    begin_info.flags = 0;
+    begin_info.pInheritanceInfo = NULL;
+
+    if ((vr = vk_procs->vkBeginCommandBuffer(vk_cmd_buffer, &begin_info)) < 0)
+    {
+        WARN("Failed to begin command buffer, vr %d.\n", vr);
+        return vr;
+    }
+
+    vk_cmd_image_barrier(vk_procs, vk_cmd_buffer,
+            VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT, VK_PIPELINE_STAGE_TRANSFER_BIT,
+            0, VK_ACCESS_TRANSFER_WRITE_BIT,
+            VK_IMAGE_LAYOUT_UNDEFINED, VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL, vk_dst_image);
+
+    blit.srcSubresource.aspectMask = VK_IMAGE_ASPECT_COLOR_BIT;
+    blit.srcSubresource.mipLevel = 0;
+    blit.srcSubresource.baseArrayLayer = 0;
+    blit.srcSubresource.layerCount = 1;
+    blit.srcOffsets[0].x = 0;
+    blit.srcOffsets[0].y = 0;
+    blit.srcOffsets[0].z = 0;
+    blit.srcOffsets[1].x = swapchain->desc.Width;
+    blit.srcOffsets[1].y = swapchain->desc.Height;
+    blit.srcOffsets[1].z = 1;
+    blit.dstSubresource = blit.srcSubresource;
+    blit.dstOffsets[0].x = 0;
+    blit.dstOffsets[0].y = 0;
+    blit.dstOffsets[0].z = 0;
+    if (swapchain->desc.Scaling == DXGI_SCALING_NONE)
+    {
+        blit.srcOffsets[1].x = min(swapchain->vk_swapchain_width, blit.srcOffsets[1].x);
+        blit.srcOffsets[1].y = min(swapchain->vk_swapchain_height, blit.srcOffsets[1].y);
+        blit.dstOffsets[1].x = blit.srcOffsets[1].x;
+        blit.dstOffsets[1].y = blit.srcOffsets[1].y;
+    }
+    else
+    {
+        /* FIXME: handle DXGI_SCALING_ASPECT_RATIO_STRETCH. */
+        blit.dstOffsets[1].x = swapchain->vk_swapchain_width;
+        blit.dstOffsets[1].y = swapchain->vk_swapchain_height;
+    }
+    blit.dstOffsets[1].z = 1;
+
+    vk_procs->vkCmdBlitImage(vk_cmd_buffer,
```

Comment:
We should rewrite this in the future to use an actual graphics pipleine since this frequently produces validation errors. This is good enough for now, but I'd like to see a `TODO` comment here.

Reply:
Resolved locally.